docs/pages/indexing.md

```diff
@@ -1,18 +1,325 @@
-with mixtures of:
-* integer scalars
-* [seq()](\ref af::seq) representing a linear sequence
-* [end](\ref af::end) representing the last element of a dimension
-* [span](\ref af::span) representing the entire dimension
+Indexing in ArrayFire is a powerful but easy to abuse feature of the af::array
+class. This feature allows you to reference or copy subsections of a larger array
+and perform operations on only a subset of elements.
+
+Indexing in ArrayFire can be performed using the parenthesis operator or one of
+the member functions of the af::array class. These functions allow you to
+reference one or a range of elements from the original array.
+
+Here we will demonstrate some of the ways you can use indexing in ArrayFire and
+discuss ways to minimize the memory and performance impact of these operations.
+
+Lets start by creating a new 4x4 matrix of floating point numbers:
+
+\snippet test/index.cpp index_tutorial_1
+
+The ArrayFire library is column-major so the resulting A array will
+look like this:
+
+\f[
+\begin{bmatrix}
+    0 & 4 & 8 & 12 \\
+    1 & 5 & 9 & 13 \\
+    2 & 6 & 10 & 14 \\
+    3 & 7 & 11 & 15
+\end{bmatrix}
+\f]
+
+This is a two dimensional array so we can access the first element of this
+matrix by passing `0,0` into the parenthesis operator of the af::array.
+
+\snippet test/index.cpp index_tutorial_first_element
+
+\f[ A(2, 3) = [ 14 ] \f]
+
+We can also access the array using linear indexing by passing in one value. Here
+we are accessing the fifth element of the array.
+
+\snippet test/index.cpp index_tutorial_fifth_element
+
+\f[ A(5) = [ 5 ] \f]
+
+\note Normally you want to avoid accessing individual elements of the array like this
+for performance reasons.
+
+Negative indexing values will index from the end of the array. For example, the
+value negative one and negative two(-2) will return the last and second to last
+element of the array. ArrayFire provides the end alias for this which also
+allows you to index the last element of the array.
+
+\snippet test/index.cpp index_tutorial_negative_indexing
+
+## Indexing slices and subarrays
+
+You can access regions of the array via the af::seq and af::span objects. The
+span objects allows you to select the entire set of elements across a particular
+dimension/axis of an array. For example, we can select the third column of the
+array by passing span as the first argument and 2 as the second argument to the
+parenthesis operator.
+
+\snippet test/index.cpp index_tutorial_third_column
+
+\f[
+A(span, 2) =
+\begin{bmatrix}
+    8 \\
+    9 \\
+    10 \\
+    11
+\end{bmatrix}
+\f]
+
+You can read that as saying that you want all values across the first dimension,
+but only from index 2 of the second dimension.
+
+You can access the second rows by passing (1, span) to the array
+
+\snippet test/index.cpp index_tutorial_second_row
+
+\f[ A(1, span) = [ 1, 5, 9, 13 ] \f]
+
+You can use the af::seq (short for sequence) object defines a range of values
```

Comment:
`... af::seq object defines a range of values ... ` 
->
 `... af::seq object to define a range ... ` ?